### DISCUSSION TITLE
InsertAllowed = ModifyAllowed = DeleteAllowed = false => Editable = false

### DISCUSSION THREAD

**Discussion by jhoek** [StefanMaron/BusinessCentral.LinterCop#428] [New Rule]
Would this perhaps be a rule that you're willing to consider for LinterCop? 

If a page has InsertAllowed = false, ModifyAllowed = false and DeleteAllowed = false, that combination of properties is functionally equivalent to Editable = false, but the latter more clearly expresses the developer's intent.

**Comment by Arthurvdv** [comment]
Good idea, I'll have a look into it

**Comment by pri-kise** [comment]
Currently we use explicit those properties to ensure that PageExtensions don't change our intent.

Editable can be changed by extensions, but the properties InsertAllowed and so on cannot be changed.

Additionally I think it's necessary to set InsertAllowed to false on API Pages, that you don't have to set DelayedInsert to true.

**Comment by Arthurvdv** [comment]
Thank you for that insight @pri-kise, I didn't realize that you could change the `Editable` property with an page extensions.

I personally didn't encounter any use cases where the page has `Editable = false` and an extension would like to change this to `true`. So I could use some help here on how to go forward on this.

We could make the rule, in case the `Extensible = false` property also is set, then raise the rule as described here. Let me show this with some examples:

```AL
page 50100 MyPage
{
    // Raise a rule to adapt to Editable = false;
    InsertAllowed = false;
    ModifyAllowed = false;
    DeleteAllowed = false;
    Extensible = false;
}
```

```AL
page 50100 MyPage
{
    // Do not raise a rule (this prohibits an page extension to change these properties)
    InsertAllowed = false;
    ModifyAllowed = false;
    DeleteAllowed = false;
    Extensible = true; // The default is true, so not explicit declaring this property here is the same
}
```

```AL
page 50100 MyPage
{
    // Do not raise a rule (page can not be extended)
    Editable = false;
    Extensible = false;
}
```

```AL
page 50100 MyPage
{
    // Raise a rule to adapt to Insert/Modified/Delete to prevent an page extension to change these properties
    Editable = false;
    Extensible = true;
}
```

In combination with [LC0034 - The property Extensible should be explicitly set for public objects.](https://github.com/StefanMaron/BusinessCentral.LinterCop/wiki/LC0034) this rule could make sense.

But to be honest I'm doubting a bit here, because of the default true value of the `Extensible` property, this rule wil in most cases don't show up and defeats the purpose?

The rule would even do the opposite I think; Modifying from Editable to Allow (to prevent a possible page extension to change the behavior).

**Comment by rvanbekkum** [comment]
I have ran into something that was very similar:

In my scenario the page had `DeleteAllowed = true;` and `Editable = false;`
The `Editable` property wins (deletion is not allowed) and so the `DeleteAllowed = true;` property does not have any effect.

I think that this example could still be a _separate_ rule which could raise a diagnostic that informs the developer that setting `Editable = false;` contradicts the setting of `Delete/Modify/InsertAllowed = true;`

**Comment by Pazzakara** [comment]
Having the 3 properties (InsertAllowed, DeleteAllowed, ModifyAllowed) set to false is **NOT** the same like Editable = false.
The 3 properies only apply to fields that show fields of the source table record. The Editable property applies to all fields, including fields with a global variable as source expression.
[ExamplePage.txt](https://github.com/StefanMaron/BusinessCentral.LinterCop/files/14448686/ExamplePage.txt)
I'd aggree to **rvanbekkum** and say it could be useful to have a rule like: if Editable is set to false then the other 3 properties are redundant (i.e. they should not be set, neither to true nor to false).

**Comment by jwikman** [comment]
One more thing to consider, even if you set `Editable = false`, some system actions (new/delete) are visible (but disabled) if you do not set them explicit to `false`.
For that reason, we have several pages with:
```
    DeleteAllowed = false;
    Editable = false;
    InsertAllowed = false;
```
I see that we also have `ModifyAllowed = false` together with `Editable = false;` in some listparts, but unsure if that was really needed. I don't have the time to deep dive into it right now, but wanted to let you know about this now anyway, so you do not raise a warning for these scenarios (even though I do not see that happen in your examples). 🙂
